### ISSUE TITLE
BaseMpscLinkedArrayQueue.peek() causes 100% CPU util

### ISSUE THREAD

**Issue by kent-b** [JCTools/JCTools#379]
After running our netty-based service for minutes to hours, eventually one thread will run in a loop and consume 100% of a core on an EC2 instance.

This is the stack trace:
```
Info-executor-11" #50 daemon prio=5 os_prio=0 cpu=353928278.59ms elapsed=594662.06s tid=0x00007fe0a400dfc0 nid=0xb09 runnable  [0x00007fe05d6ed000]
   java.lang.Thread.State: RUNNABLE
	at io.netty.util.internal.shaded.org.jctools.queues.BaseMpscLinkedArrayQueue.peek(BaseMpscLinkedArrayQueue.java:382)
	at io.netty.util.internal.shaded.org.jctools.queues.MpscUnboundedArrayQueue.peek(MpscUnboundedArrayQueue.java:23)
	at co.cask.http.internal.NonStickyEventExecutorGroup$NonStickyOrderedEventExecutor.run(NonStickyEventExecutorGroup.java:286)
	at io.netty.util.concurrent.UnorderedThreadPoolEventExecutor$NonNotifyRunnable.run(UnorderedThreadPoolEventExecutor.java:292)
	at java.util.concurrent.Executors$RunnableAdapter.call(java.base@17.0.8/Executors.java:539)
	at java.util.concurrent.FutureTask.run(java.base@17.0.8/FutureTask.java:264)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(java.base@17.0.8/ScheduledThreadPoolExecutor.java:304)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@17.0.8/ThreadPoolExecutor.java:1136)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@17.0.8/ThreadPoolExecutor.java:635)
	at java.lang.Thread.run(java.base@17.0.8/Thread.java:833)
```

**Netty version**
4.1.95.Final

**JCTools version**
3.1.0

**openjdk version**
"17.0.8" 2023-07-18
OpenJDK Runtime Environment (build 17.0.8+7-Ubuntu-120.04.2)
OpenJDK 64-Bit Server VM (build 17.0.8+7-Ubuntu-120.04.2, mixed mode, sharing)

**OS version (e.g. uname -a)**
Ubuntu 20.04.6 LTS
kernel 5.15.0-1036-aws x86_64

**Comment by ceragon** [comment]
First, I think your netty version is not 4.1.95.Final, but a version before 4.1.42.Final. Second, the issue has been fixed in version 4.1.42.Final, and the related issue is [Fix possible infinite while loop in NonStickyEventExecutorGroup due misusage of MPSC queue](http://www.bing.com/translator/?ref=TThis&text=&from=zh-cn&to=en).


- According to your call stack, NonStickyEventExecutorGroup called the peek method of MpscUnboundedArrayQueue

```
at io.netty.util.internal.shaded.org.jctools.queues.MpscUnboundedArrayQueue.peek(MpscUnboundedArrayQueue.java:23)
at co.cask.http.internal.NonStickyEventExecutorGroup$NonStickyOrderedEventExecutor.run(NonStickyEventExecutorGroup.java:286)
```

- However, in 4.1.95.Final, there is no place in the NonStickyEventExecutorGroup class that calls the peek method.
But I checked the commit history of the NonStickyEventExecutorGroup class, and the peek method was added in this commit [Fix race condition in the NonStickyEventExecutorGroup by chtyim · Pull Request #8232](https://github.com/netty/netty/pull/8232).

```
if (tasks.peek() == null || !state.compareAndSet(NONE, RUNNING)) {
	return; // done
}
```

- But due to an infinite loop problem, it was changed to the isEmpty method in this commit [Fix possible infinite while loop in NonStickyEventExecutorGroup due misusage of MPSC queue](http://www.bing.com/translator/?ref=TThis&text=&from=zh-cn&to=en).
```
if (tasks.isEmpty() || !state.compareAndSet(NONE, RUNNING)) {  
    // Only set executingThread to null if no other thread did update it yet.  
    executingThread.compareAndSet(current, null);  
    return; // done  
}
```

Therefore, I think you can solve your problem by upgrading your netty version to the real 4.1.95.Final version.

**Comment by franz1981** [comment]
Thanks @ceragon yep is it accurate, I think I can close this as explained
